pkg/ccl/streamingccl/logical/logical_replication_writer_processor.go

```diff
@@ -802,54 +807,86 @@ func (lrw *logicalReplicationWriterProcessor) flushBatch(
-		// thus not part of the rangefeed datum.
-		if col.IsVirtual() {
-			continue
+func makeInsertQueries(
+	fqTableName string, td catalog.TableDescriptor,
+) (map[catid.FamilyID]string, error) {
+	queries := make(map[catid.FamilyID]string, td.NumFamilies())
+
+	if err := td.ForeachFamily(func(family *descpb.ColumnFamilyDescriptor) error {
+		var columnNames strings.Builder
+		var valueStrings strings.Builder
+		var onConflictUpdateClause strings.Builder
+		argIdx := 1
+		seenIds := make(map[catid.ColumnID]struct{})
+		addColumn := func(colName string, colID catid.ColumnID) {
+			// We will set crdb_internal_origin_timestamp ourselves from the MVCC timestamp of the incoming datum.
+			// We should never see this on the rangefeed as a non-null value as that would imply we've looped data around.
+			if colName == "crdb_internal_origin_timestamp" {
+				return
+			}
+			if _, seen := seenIds[colID]; seen {
+				return
+			}
+
+			if argIdx == 1 {
+				columnNames.WriteString(colName)
+				fmt.Fprintf(&valueStrings, "$%d", argIdx)
+				fmt.Fprintf(&onConflictUpdateClause, "%s = $%d", colName, argIdx)
+			} else {
+				fmt.Fprintf(&columnNames, ", %s", colName)
+				fmt.Fprintf(&valueStrings, ", $%d", argIdx)
+				fmt.Fprintf(&onConflictUpdateClause, ",\n%s = $%d", colName, argIdx)
+			}
+			seenIds[colID] = struct{}{}
+			argIdx++
 		}
-		// We will set crdb_internal_origin_timestamp ourselves from the MVCC timestamp of the incoming datum.
-		// We should never see this on the rangefeed as a non-null value as that would imply we've looped data around.
-		if col.GetName() == "crdb_internal_origin_timestamp" {
-			continue
+
+		publicColumns := td.PublicColumns()
+		colOrd := catalog.ColumnIDToOrdinalMap(publicColumns)
+		primaryIndex := td.GetPrimaryIndex()
+
+		for i := 0; i < primaryIndex.NumKeyColumns(); i++ {
+			for i := 0; i < primaryIndex.NumKeyColumns(); i++ {
+				ord, ok := colOrd.Get(primaryIndex.GetKeyColumnID(i))
+				if !ok {
+					return errors.AssertionFailedf("expected to find column %d", ord)
+				}
+				col := publicColumns[ord]
+				addColumn(col.GetName(), col.GetID())
+
+			}
 		}
-		if argIdx == 1 {
-			columnNames.WriteString(col.GetName())
-			fmt.Fprintf(&valueStrings, "$%d", argIdx)
-			fmt.Fprintf(&onConflictUpdateClause, "%s = $%d", col.GetName(), argIdx)
-		} else {
-			fmt.Fprintf(&columnNames, ", %s", col.GetName())
-			fmt.Fprintf(&valueStrings, ", $%d", argIdx)
-			fmt.Fprintf(&onConflictUpdateClause, ",\n%s = $%d", col.GetName(), argIdx)
+
+		for i, colName := range family.ColumnNames {
+			addColumn(colName, family.ColumnIDs[i])
 		}
-		argIdx++
-	}
-	originTSIdx := argIdx
-	baseQuery := `
+
+		originTSIdx := argIdx
+		baseQuery := `
 INSERT INTO %s (%s, crdb_internal_origin_timestamp)
 VALUES (%s, $%d)
 ON CONFLICT ON CONSTRAINT %s
 DO UPDATE SET
 %s,
 crdb_internal_origin_timestamp=$%[4]d
-WHERE (%[1]s.crdb_internal_mvcc_timestamp < $%[4]d
+WHERE (%[1]s.crdb_internal_mvcc_timestamp <= $%[4]d
```

Comment:
heh, tie goes to the runner i guess?

Reply:
Yeah. The real reason for this change is what we chatted about around a single update that updates both column families. If we didn't have the equality, we would end up dropping one of the updates.